test/Reader.hs

```diff
@@ -11,53 +12,60 @@ import qualified Control.Monad.Trans.Reader as ReaderT
 import qualified Control.Monad.Trans.RWS.Lazy as LazyRWST
 import qualified Control.Monad.Trans.RWS.Strict as StrictRWST
 import Data.Function ((&))
-import Data.Functor.Identity (Identity)
 import Gen
+import GHC.Generics ((:.:)(..))
 import qualified Monad
 import qualified MonadFix
 import Test.Tasty
 import Test.Tasty.Hedgehog
 
 tests :: TestTree
 tests = testGroup "Reader"
-  [ testGroup "ReaderC"       $
+  [ testGroup "ReaderC" $
     [ testMonad
     , testMonadFix
     , testReader
-    ] >>= ($ RunR (uncurry ReaderC.runReader))
-  , testGroup "(->)"          $ testReader (RunR (uncurry (fmap PureC . (&))))
-  , testGroup "ReaderT"       $ testReader (RunR (uncurry (flip ReaderT.runReaderT)))
-  , testGroup "RWST (Lazy)"   $ testReader (RunR (uncurry (runRWST LazyRWST.runRWST)))
-  , testGroup "RWST (Strict)" $ testReader (RunR (uncurry (runRWST StrictRWST.runRWST)))
+    ] >>= ($ runR (uncurry ReaderC.runReader . lower))
+  , testGroup "(->)"          $ testReader (runR (uncurry (fmap pure . (&))            . lower))
+  , testGroup "ReaderT"       $ testReader (runR (uncurry (flip ReaderT.runReaderT)    . lower))
+  , testGroup "RWST (Lazy)"   $ testReader (runR (uncurry (runRWST LazyRWST.runRWST)   . lower))
+  , testGroup "RWST (Strict)" $ testReader (runR (uncurry (runRWST StrictRWST.runRWST) . lower))
   ] where
-  testMonad    run = Monad.test    (m (gen r)) a b c (atom "(,)" (,) <*> r <*> unit) run
-  testMonadFix run = MonadFix.test (m (gen r)) a b   (atom "(,)" (,) <*> r <*> unit) run
-  testReader   run = Reader.test r (m (gen r)) a                                     run
+  testMonad    run = Monad.test    (m (gen0 r) (genN r)) a b c (Comp1 <$> (identity <*> (pair <*> r <*> unit))) run
+  testMonadFix run = MonadFix.test (m (gen0 r) (genN r)) a b   (Comp1 <$> (identity <*> (pair <*> r <*> unit))) run
+  testReader   run = Reader.test r (m (gen0 r) (genN r)) a                (identity <*>                 unit)   run
   runRWST f r m = (\ (a, _, ()) -> a) <$> f m r r
+  lower = runIdentity . unComp1
 
 
-gen
-  :: forall r m sig
+gen0
+  :: forall r m a sig
   .  (Has (Reader r) sig m, Arg r, Show r, Vary r)
   => Gen r
+  -> Gen a
+  -> [Gen (m a)]
+gen0 _ a = [ label "asks" (asks @r) <*> fn a ]
+
+genN
+  :: (Has (Reader r) sig m, Arg r, Show r, Vary r)
+  => Gen r
   -> GenM m
-  -> GenM m
-gen r m a = choice
-  [ label "asks" (asks @r) <*> fn a
-  , label "local" local <*> fn r <*> m a
-  ]
+  -> Gen a
+  -> [Gen (m a)]
+genN r m a = [ label "local" local <*> fn r <*> m a ]
 
 
 test
-  :: (Has (Reader r) sig m, Arg r, Eq a, Show a, Show r, Vary r)
+  :: (Has (Reader r) sig m, Arg r, Eq a, Show a, Show r, Vary r, Functor f)
   => Gen r
   -> GenM m
   -> Gen a
-  -> RunC r Identity m
+  -> Gen (f ())
+  -> Run (f :.: (,) r) Identity m
```

Comment:
`Reader` already had input state, so we add an extra layer with `:.:`. Note that we have separate generators for the `f` part and the `(,) r` part, because we can’t in general extract `r` from `f (r, ())` without knowledge about what `f` is.